### ISSUE TITLE
Login error

### ISSUE THREAD

**Issue by ac4bfg** [instaloader/instaloader#1758]
Hi,

I got error `KeyError: "name='csrftoken', domain=None, path=None"`,  when trying to login

instaloder ver : 4.9.5

**Comment by GianlucaDiFranco** [comment]
Hi, 
I have the same problem.

**Comment by tbrtje** [comment]
Yeah, me too.
This only happens with one account I have

**Comment by FinnMal** [comment]
+1

**Comment by carlosevi94** [comment]
Same here!

**Comment by IL2C** [comment]
I'm not getting this error. Can someone who is print the response text from the post request?
https://github.com/instaloader/instaloader/blob/385c6c8a353458b8e1f5d1eacd98a714a41e9fc0/instaloader/instaloadercontext.py#L222

**Comment by AyluinReymaer** [comment]
> I'm not getting this error. Can someone who is print the response text from the post request?
> 
> https://github.com/instaloader/instaloader/blob/385c6c8a353458b8e1f5d1eacd98a714a41e9fc0/instaloader/instaloadercontext.py#L222

I've printed the response like so:

``` python
resp_json = login.json()
print(resp_json, flush=True)
```

printing just the `login` variable will just give me status `200`.

This is the output (redacted the userId):

```
{'user': True, 'userId': '3765*****', 'authenticated': True, 'oneTapPrompt': True, 'status': 'ok'}
```

One thing to note, today instagram kicked me out (not temp ban, just log out) and had to login again.

Apparently, the login is now different. I had to login with username and password (the usual) but then, they asked me to confirm my username and click next, after doing so, they gave me 3 options to choose for 2fa which I chose SMS and was able to login without issues. Not sure if it is relevant.

**Comment by AyluinReymaer** [comment]
It is also good to note that instaloader **does** indeed login to instagram.

If I go to the login activity in instagram, I can see that instaloader was able to successfully login. It is just saving the session in a session file that this seems to be failing.

**Comment by ulrichhave** [comment]
I have this problem with 2 different accounts. Experienced it in both version 4.9.3 and 4.9.5 - any solutions?

**Comment by evanmiltenburg** [comment]
I have the same issue, but as @AyluinReymaer noted: Instaloader does log in.

**Comment by Coco1923** [comment]
Same issue

**Comment by evanmiltenburg** [comment]
I think I have a workaround: 
* Use Firefox to login to Instagram
* Use the script from [here](https://instaloader.github.io/troubleshooting.html#login-error) to copy the session file from Firefox to the Instaloader path.
* Use the following code to log in:

```python
USERNAME = "your_username"

import instaloader

L = instaloader.Instaloader()
L.load_session_from_file(USERNAME)
```

**Comment by ulrichhave** [comment]
@evanmiltenburg thanks, it works!

**Comment by Zorpubis** [comment]
> I think I have a workaround:
> 
> * Use Firefox to login to Instagram
> * Use the script from [here](https://instaloader.github.io/troubleshooting.html#login-error) to copy the session file from Firefox to the Instaloader path.
> * Use the following code to log in:
> 
> ```python
> USERNAME = "your_username"
> 
> import instaloader
> 
> L = instaloader.Instaloader()
> L.load_session_from_file(USERNAME)
> ```

Downloading the session file is no problem but this still doesn't work : 

- logging in this way seems to fail but doesn't throw any error. It's until you execute the command "instaloader -l USERNAME" that then the password is required (new ?), and then get the original error above
- trying to log in in without session file will throw you the error right away

**Comment by ErikUden** [comment]
Same issue

**Comment by ErikUden** [comment]
> > I think I have a workaround:
> > 
> > * Use Firefox to login to Instagram
> > * Use the script from [here](https://instaloader.github.io/troubleshooting.html#login-error) to copy the session file from Firefox to the Instaloader path.
> > * Use the following code to log in:
> > 
> > ```python
> > USERNAME = "your_username"
> > 
> > import instaloader
> > 
> > L = instaloader.Instaloader()
> > L.load_session_from_file(USERNAME)
> > ```
> 
> Downloading the session file is no problem but this still doesn't work :
> 
>     * logging in this way seems to fail but doesn't throw any error. It's until you execute the command "instaloader -l USERNAME" that then the password is required (new ?), and then get the original error above
> 
>     * trying to log in without session file will throw you the error right away

Trying to follow this method I even get the following error:
```
Using cookies from C:\Users\MYUSERNAME/AppData/Roaming/Mozilla/Firefox/Profiles\SESSIONFILENAME.default-release\cookies.sqlite.
Not logged in. Are you logged in successfully in Firefox?
```
I am currently logged into Instagram. Not sure what else to do?
Wasn't there a browser plugin that could download Sessions/Cookies?

**Comment by ErikUden** [comment]
> 



> I think I have a workaround:
> 
>     * Use Firefox to login to Instagram
> 
>     * Use the script from [here](https://instaloader.github.io/troubleshooting.html#login-error) to copy the session file from Firefox to the Instaloader path.
> 
>     * Use the following code to log in:
> 
> 
> ```python
> USERNAME = "your_username"
> 
> import instaloader
> 
> L = instaloader.Instaloader()
> L.load_session_from_file(USERNAME)
> ```

This is how I fixed it:

HOW TO FIX `KeyError: "name='csrftoken', domain=None, path=None"` BUG
(assuming you have Instaloader already installed)

# STEP 1:
*Follow these steps:*
(based on: https://instaloader.github.io/troubleshooting.html#login-error )

**A.** Download the script: Example 615_import_firefox_session.py
https://raw.githubusercontent.com/instaloader/instaloader/master/docs/codesnippets/615_import_firefox_session.py
**B.** Open Firefox and go to the "Privacy & Security" settings (about:preferences#privacy)
**C.** Change "History"-Settings from "Use custom settings for history" -> "Always use custom browsing mode" to "*Remember history*". 
> "Firefox must restart to disable this feature." -> "Restart Firefox now"
**D.** Open Firefox again and login to Instagram in Firefox with whatever account you then wish to use for Instaloaders.

# STEP 2: 
**A.** Open CMD as administrator and go to wherever you downloaded the script to
**B.** Execute `python 615_import_firefox_session.py`
> Your Instagram-session should be imported now and usable!

# STEP 3:
Problem solved, now execute any Instaloader command using the "--login=usernameFromSession" argument

**Comment by FinnMal** [comment]
Not working for me on MacOS. I can see the "csrftoken" in the sql file, but after running the script, the error shows up again. 

I've also tried to do the login directly after the "instaloader.save_session_to_file(sessionfile)" in the script -> no difference, same error.

**Comment by NicotherZ** [comment]
> > 
> 
> > I think I have a workaround:
> > ```
> > * Use Firefox to login to Instagram
> > 
> > * Use the script from [here](https://instaloader.github.io/troubleshooting.html#login-error) to copy the session file from Firefox to the Instaloader path.
> > 
> > * Use the following code to log in:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```python
> > USERNAME = "your_username"
> > 
> > import instaloader
> > 
> > L = instaloader.Instaloader()
> > L.load_session_from_file(USERNAME)
> > ```
> 
> This is how I fixed it:
> 
> HOW TO FIX `KeyError: "name='csrftoken', domain=None, path=None"` BUG (assuming you have Instaloader already installed)
> # STEP 1:
> 
> _Follow these steps:_ (based on: https://instaloader.github.io/troubleshooting.html#login-error )
> 
> **A.** Download the script: Example 615_import_firefox_session.py https://raw.githubusercontent.com/instaloader/instaloader/master/docs/codesnippets/615_import_firefox_session.py **B.** Open Firefox and go to the "Privacy & Security" settings (about:preferences#privacy) **C.** Change "History"-Settings from "Use custom settings for history" -> "Always use custom browsing mode" to "_Remember history_".
> 
> > "Firefox must restart to disable this feature." -> "Restart Firefox now"
> > **D.** Open Firefox again and login to Instagram in Firefox with whatever account you then wish to use for Instaloaders.
> 
> # STEP 2:
> 
> **A.** Open CMD as administrator and go to wherever you downloaded the script to **B.** Execute `python 615_import_firefox_session.py`
> 
> > Your Instagram-session should be imported now and usable!
> 
> # STEP 3:
> 
> Problem solved, now execute any Instaloader command using the "--login=usernameFromSession" argument

Working one time and after that, problem coming back, have to do it again and again each time i want use instaloader :/

**Comment by Zorpubis** [comment]
> > 
> 
> > I think I have a workaround:
> > ```
> > * Use Firefox to login to Instagram
> > 
> > * Use the script from [here](https://instaloader.github.io/troubleshooting.html#login-error) to copy the session file from Firefox to the Instaloader path.
> > 
> > * Use the following code to log in:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```python
> > USERNAME = "your_username"
> > 
> > import instaloader
> > 
> > L = instaloader.Instaloader()
> > L.load_session_from_file(USERNAME)
> > ```
> 
> This is how I fixed it:
> 
> HOW TO FIX `KeyError: "name='csrftoken', domain=None, path=None"` BUG (assuming you have Instaloader already installed)
> 
> # STEP 1:
> _Follow these steps:_ (based on: https://instaloader.github.io/troubleshooting.html#login-error )
> 
> **A.** Download the script: Example 615_import_firefox_session.py https://raw.githubusercontent.com/instaloader/instaloader/master/docs/codesnippets/615_import_firefox_session.py **B.** Open Firefox and go to the "Privacy & Security" settings (about:preferences#privacy) **C.** Change "History"-Settings from "Use custom settings for history" -> "Always use custom browsing mode" to "_Remember history_".
> 
> > "Firefox must restart to disable this feature." -> "Restart Firefox now"
> > **D.** Open Firefox again and login to Instagram in Firefox with whatever account you then wish to use for Instaloaders.
> 
> # STEP 2:
> **A.** Open CMD as administrator and go to wherever you downloaded the script to **B.** Execute `python 615_import_firefox_session.py`
> 
> > Your Instagram-session should be imported now and usable!
> 
> # STEP 3:
> Problem solved, now execute any Instaloader command using the "--login=usernameFromSession" argument

Yup, same thing

No problem with configuring Firefox and getting the session file but : 

`instaloader --login=usernameFromSession`

will still ask for a password. 

Was logging in with a session file asking for a password before ? I don't recall so. Anyone ? 

And even if you type it, it will result in the original error thrown away when trying to log in with credentials given :

`KeyError: "name='csrftoken', domain=None, path=None"`

**Comment by gusioo** [comment]
@Zorpubis you can use -f switch instead of --login.

-f use session file instead of login/pass

**Comment by Nova6K0** [comment]
I am preparing a file, to upload (which will be an explanation of the problems we are having, and why. But it is an unfinished version. Although enough to explain, certain things. The reason for not finishing it quickly, is that my English It is very bad, and use the Google Translate a lot, which is not very good either. And as I am Spanish, well ...

Greetings.

**Comment by IL2C** [comment]
Replacing the default user agent with any from [this](https://developers.whatismybrowser.com/useragents/explore/software_name/instagram/) list resolves this issue.
https://github.com/instaloader/instaloader/blob/ffb0aea5eb79b58ebad1eee57ad097bc2aba0689/instaloader/instaloadercontext.py#L34

**Comment by evanmiltenburg** [comment]
Of you have a working version, can you perhaps submit a pull request?

**Comment by yah00078** [comment]
Many thanks for workarounds or corrections, I’ll try them and keep in touch ! Thanks for all this work !

Vincent,

Envoyé de mon SmartPhone .

**Comment by NicotherZ** [comment]
I try : 615_import_firefox_session.py but it close instantly and does nothing.
Try with IDLE and red message appear : 

"SystemExit: Instaloader not found.
  pip install [--user] instaloader"

I use windows version.

replace by -f don't download in best quality. 
And don't know what to do with the IL2C's fix (im a noob in coding ^^)

Someone can help me please ?

**Comment by evanmiltenburg** [comment]
Thanks @IL2C! I have tested your code and can confirm logging in does not produce an error anymore. So for those of you who cannot wait to use Instaloader again:
```
pip uninstall instaloader
pip install git+https://github.com/IL2C/instaloader.git@login-fix
```

(Of course you should always double-check when you're installing code from someone else's repository. But at the time of posting the repo was exactly 1 commit ahead of the Instaloader main repo, with only the headers changed.)

**Comment by IL2C** [comment]
> While IL2C's patch fixes the error, it now produces another error: _"Fatal error: Login error: Unexpected response, this might indicate a blocked IP."_
> 
> (And just to be clear, my IP is not blocked. I'm successfully logging in with a browser on the same computer and same account, and other accounts also work from this IP.)

Could you share your code and the full stack trace?

**Comment by IL2C** [comment]
> While IL2C's patch fixes the error, it now produces another error: _"Fatal error: Login error: Unexpected response, this might indicate a blocked IP."_
> 
> (And just to be clear, my IP is not blocked. I'm successfully logging in with a browser on the same computer and same account, and other accounts also work from this IP.)

#1635

**Comment by evdmolen** [comment]
Same problem here. Created a new account on insta, did a few tests on chrome. Tried my script and i was required to login. With logging in on the usual way i have the error mentioned above. Had to install firefox to use the fix. Testing my script and it did work but half hour later my brand-new account was banned. 

So terribly annoying..

**Comment by Shinoby92** [comment]
> Replacing the default user agent with any from [this](https://developers.whatismybrowser.com/useragents/explore/software_name/instagram/) list resolves this issue.
> 
> https://github.com/instaloader/instaloader/blob/ffb0aea5eb79b58ebad1eee57ad097bc2aba0689/instaloader/instaloadercontext.py#L34
> 
> From what I've gathered, using a user agent that doesn't correlate Instagram software won't access the csrftoken cookie when sending the post request to login.
> 
> https://github.com/instaloader/instaloader/blob/3cc29a4ceb3ff4cd04a73dd4b20979b03128f454/instaloader/instaloadercontext.py#L222

this fixed the issue
pull up request submit
Thank you

**Comment by francescoavant** [comment]
Hi, same problem.
The solutions proposed above are not working for me. I've tried with 5 different accounts

**Comment by Shinoby92** [comment]
I got 3 accounts banned today. Create a new one and try it with that. Seems that they getting flagged after wrong using

**Comment by eldepor** [comment]
> Thanks @IL2C! I have tested your code and can confirm logging in does not produce an error anymore. So for those of you who cannot wait to use Instaloader again:
> 
> ```
> pip uninstall instaloader
> pip install git+https://github.com/IL2C/instaloader.git@login-fix
> ```
> 
> (Of course you should always double-check when you're installing code from someone else's repository. But at the time of posting the repo was exactly 1 commit ahead of the Instaloader main repo, with only the headers changed.)

This worked beautifully for me. Thanks.

**Comment by Maxklos** [comment]
> Thanks @IL2C! I have tested your code and can confirm logging in does not produce an error anymore. So for those of you who cannot wait to use Instaloader again:
> 
> ```
> pip uninstall instaloader
> pip install git+https://github.com/IL2C/instaloader.git@login-fix
> ```
> 
> (Of course you should always double-check when you're installing code from someone else's repository. But at the time of posting the repo was exactly 1 commit ahead of the Instaloader main repo, with only the headers changed.)

Works for me with no issues. Thanks!

**Comment by mrousse83** [comment]
Hello,

I have the same problem with 4.9.5 (windows standalone).

Do you have a solution for this binary version ?

Thanks,
Mathieu

**Comment by b5kaa** [comment]
Hello, I started receiving this error around 10 days ago after using instaloader for quite a while. The thing is I can't login to either of my instagram on any of my browsers, remote desktops et.c. I tried clearing cookies, trying different VPNs. The account are still operating as I can switch to them on my mobile device. But I can't login there in the browser version on the same device.

**Comment by b5kaa** [comment]
> > 
> 
> > I think I have a workaround:
> > ```
> > * Use Firefox to login to Instagram
> > 
> > * Use the script from [here](https://instaloader.github.io/troubleshooting.html#login-error) to copy the session file from Firefox to the Instaloader path.
> > 
> > * Use the following code to log in:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```python
> > USERNAME = "your_username"
> > 
> > import instaloader
> > 
> > L = instaloader.Instaloader()
> > L.load_session_from_file(USERNAME)
> > ```
> 
> This is how I fixed it:
> 
> HOW TO FIX `KeyError: "name='csrftoken', domain=None, path=None"` BUG (assuming you have Instaloader already installed)
> # STEP 1:
> 
> _Follow these steps:_ (based on: https://instaloader.github.io/troubleshooting.html#login-error )
> 
> **A.** Download the script: Example 615_import_firefox_session.py https://raw.githubusercontent.com/instaloader/instaloader/master/docs/codesnippets/615_import_firefox_session.py **B.** Open Firefox and go to the "Privacy & Security" settings (about:preferences#privacy) **C.** Change "History"-Settings from "Use custom settings for history" -> "Always use custom browsing mode" to "_Remember history_".
> 
> > "Firefox must restart to disable this feature." -> "Restart Firefox now"
> > **D.** Open Firefox again and login to Instagram in Firefox with whatever account you then wish to use for Instaloaders.
> 
> # STEP 2:
> 
> **A.** Open CMD as administrator and go to wherever you downloaded the script to **B.** Execute `python 615_import_firefox_session.py`
> 
> > Your Instagram-session should be imported now and usable!
> 
> # STEP 3:
> 
> Problem solved, now execute any Instaloader command using the "--login=usernameFromSession" argument

I now encounter

> > 
> 
> > I think I have a workaround:
> > ```
> > * Use Firefox to login to Instagram
> > 
> > * Use the script from [here](https://instaloader.github.io/troubleshooting.html#login-error) to copy the session file from Firefox to the Instaloader path.
> > 
> > * Use the following code to log in:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```python
> > USERNAME = "your_username"
> > 
> > import instaloader
> > 
> > L = instaloader.Instaloader()
> > L.load_session_from_file(USERNAME)
> > ```
> 
> This is how I fixed it:
> 
> HOW TO FIX `KeyError: "name='csrftoken', domain=None, path=None"` BUG (assuming you have Instaloader already installed)
> # STEP 1:
> 
> _Follow these steps:_ (based on: https://instaloader.github.io/troubleshooting.html#login-error )
> 
> **A.** Download the script: Example 615_import_firefox_session.py https://raw.githubusercontent.com/instaloader/instaloader/master/docs/codesnippets/615_import_firefox_session.py **B.** Open Firefox and go to the "Privacy & Security" settings (about:preferences#privacy) **C.** Change "History"-Settings from "Use custom settings for history" -> "Always use custom browsing mode" to "_Remember history_".
> 
> > "Firefox must restart to disable this feature." -> "Restart Firefox now"
> > **D.** Open Firefox again and login to Instagram in Firefox with whatever account you then wish to use for Instaloaders.
> 
> # STEP 2:
> 
> **A.** Open CMD as administrator and go to wherever you downloaded the script to **B.** Execute `python 615_import_firefox_session.py`
> 
> > Your Instagram-session should be imported now and usable!
> 
> # STEP 3:
> 
> Problem solved, now execute any Instaloader command using the "--login=usernameFromSession" argument

After doing this it worked for me, but when I tried switching things to normal I get CSRF token missing or incorrect in my browser, how can I solve it?

**Comment by francescoavant** [comment]
> 

this solution doesn't work for me.
I have tried the other proposed solutions but they do not work

**Comment by francescoavant** [comment]
> Thanks @IL2C! I have tested your code and can confirm logging in does not produce an error anymore. So for those of you who cannot wait to use Instaloader again:
> 
> ```
> pip uninstall instaloader
> pip install git+https://github.com/IL2C/instaloader.git@login-fix
> ```
> 
> (Of course you should always double-check when you're installing code from someone else's repository. But at the time of posting the repo was exactly 1 commit ahead of the Instaloader main repo, with only the headers changed.)

this solution doesn't work for me.
I have tried the other proposed solutions but they do not work

**Comment by m-sundman** [comment]
The only thing the code does is changes the default user agent. You can override the user agent more easily with a command line parameter. In any case, I still get instant ban when I try to login using this code and/or the specified user agent.

**Comment by wdstk** [comment]
Now Instagram kicks me out even though I use  @evanmiltenburg's work-around. Anyone else?

**Comment by aandergr** [comment]
Fixed in [v4.10.1](https://github.com/instaloader/instaloader/releases/tag/v4.10.1) thanks to #2088.

**Comment by yah00078** [comment]
Thank you I'll try !
